### DISCUSSION TITLE
AS4Encryptor - Symetric Key Change

### DISCUSSION THREAD

**Discussion by oaxfreeman** [phax/phase4#289] [Q&A] [answered]
Referring to `com.helger.phase4.messaging.crypto.AS4Encryptor`:

I would like to impact the symmetric key generation type. More precisely, I have to change the generated key to `KeyUtils.getKeyGenerator(AES_128_GCM)`;

![{EDF888F0-3003-44B2-BE1E-0A3614F8D70D}](https://github.com/user-attachments/assets/6682dc9d-473a-47fd-9208-2b11d9292433)

Is `ICryptoSessionKeyProvider` and option? Please @phax point me in right direction and I'm sure we'll figure it out :)

**Comment by phax** [comment]
Yes exactly. In the `AS4CryptParams` you can set the `ICryptoSessionKeyProvider`. By default a random AES-128 key is generated. I guess it makes sense to a default solution for AES-128 GCM as well, so I did :) Will be part of the 3.0.1 release

**Reply by oaxfreeman** [reply]
I don't think it might be related but :
`Phase4CEFSender.builder()`is throwing `java.lang.ClassNotFoundException: com.helger.security.keystore.ITrustStoreDescriptor`

![image](https://github.com/user-attachments/assets/98ee6bb0-0cc1-4bbe-9f6c-c7947869fd35)

What happend in version 3.0.0 so that this error occurs?

**Reply by phax** [reply]
Please make sure, that you are using ph-commons 11.1.10 in your code - previous versions will not work.

**Reply by oaxfreeman** [reply]
It is there. Also the error is not deterministic. This time it is `IKeyStoreAndKeyDescriptor` missing, @phax is there any custom classloaders in the project?
![image](https://github.com/user-attachments/assets/dddf545f-5ec7-4496-913d-3f12bfd52196)

**Reply by phax** [reply, marked as answer]
As I can see in your image you are e.g. using "ph-collections 11.1.8" - this should also be 11.1.10
The best way to include all the same versions for dependency submodules is via the BOM import in your pom.xml.
Example for ph-commons stuff only:
```xml
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.helger.commons</groupId>
        <artifactId>ph-commons-parent-pom</artifactId>
        <version>11.1.10</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </depndencies>
  </dependencyManagement>
```

This ensures, that all submodules from the ph-commons project use the same version.
Alternatively you don't explicitly include this at all and only use the phase4 include - this should transitively pick the right version

**Reply by oaxfreeman** [reply]
Had to delete entire `.m2` folder but it worked, thx

**Reply by phax** [reply]
This should not be the case. Feel free to send me your pom.xml per mail (see my pom) and I can provide you with a proposal...

**Reply by phax** [reply]
@piotr-dajlido by rechecking the code, it seems like there is no difference in the key generator with or without GCM, as the key itself needs no padding.... The padding is only needed on the content in preparation for signing/crypting. And when you look at the code, it does exactly the same:

![grafik](https://github.com/user-attachments/assets/4b06be3e-4dcc-4230-86aa-15562580a20e)

Therefore, the constants with GCM suffix will be removed again

**Reply by oaxfreeman** [reply]
Well that might be the case why my impl is throwing an error when sending a message to the endpoint that expects GCM. Thanks for pointing this out, I will investigate further.

**Reply by phax** [reply]
Usually you define the GCM padding e.g. in the PMode leg `setX509EncryptionAlgorithm` (as in `aPModeLegSecurity.setX509EncryptionAlgorithm (ECryptoAlgorithmCrypt.AES_128_GCM);` e.g. for Peppol)

**Reply by oaxfreeman** [reply]
Well that is even more interesting, I have been pushing GCM settings through message builder setup:

```
public class CryptoSettings {
    public static void apply(Phase4CEFSender.CEFUserMessageBuilder builder) {
        builder.cryptParams().setSessionKeyProvider(MockCryptoSessionKeyProvider.INSTANCE_RANDOM_AES_128_GCM);
        builder.cryptParams().setEncryptSymmetricSessionKey(true);
        builder.cryptParams().setWSSecEncryptCustomizer(new WSSecConfig());
    }
}
```
```
class WSSecConfig implements IWSSecEncryptCustomizer {
    @Nonnull
    public WSSecEncrypt createWSSecEncrypt(@Nonnull final WSSecHeader aSecHeader)
    {
        final WSSecEncrypt aBuilder = new WSSecEncrypt (aSecHeader);
        // As the receiver MAY not have pre-configured the signing leaf certificate,
        // a BinarySecurityToken token reference MUST be used to reference the
        // signing certificate.
        aBuilder.setUserInfo("wss40");
        // Encrypted key must be contained
        aBuilder.setEncryptSymmKey (true);
        aBuilder.setSymmetricEncAlgorithm(WSConstants.AES_128_GCM);
        aBuilder.setKeyIdentifierType(WSConstants.BST_DIRECT_REFERENCE);

        return aBuilder;
    }
}
```
@phax how do you acctually configure `aPModeLegSecurity`? Domibus does it through XML file, how does Phase4 handle this config?

**Reply by phax** [reply]
In phase4 the default security configurations are encapsulated in the so called `profiles` - by chosing the profile you want to use, basic stuff like signature and encryption algorithms are set to the correct values. You chose the profiles, by including them as artifacts in your build tool.
Please remind me: what AS4 profile do you want to implement?

**Reply by oaxfreeman** [reply]
CEF

**Reply by phax** [reply]
Well than you add `phase4-profile-cef` to your dependencies and you should be good.
I heard from other sides, that in some cases they require the usage of `SubjectKeyIdentifier` instead of the `BinarySecurityToken` but that is setup dependent.
If you provide me with the exact specification you need to implement, I can provide better suggestions.

**Reply by oaxfreeman** [reply]
The main goal is to exchange message between Phase4 and Domibus. 
Phase4 act as a sender, Domibus as a receiver. 
Domibus policy configuration is presented at point 3, below. 

1. Sender logic:
```java
    Phase4CEFSender.CEFUserMessageBuilder prepareMessage(MockMessage message) {
        final IParticipantIdentifier aReceiverID = SimpleIdentifierFactory.INSTANCE.createParticipantIdentifier(
                message.metadata().getPartyInfo().toPartyType(),
                message.metadata().getPartyInfo().toPartyId());

        final IParticipantIdentifier aSenderID = SimpleIdentifierFactory.INSTANCE.createParticipantIdentifier(
                message.metadata().getPartyInfo().fromPartyType(),
                message.metadata().getPartyInfo().fromPartyId());

        Phase4CEFSender.CEFUserMessageBuilder builder = Phase4CEFSender.builder()
                .cryptoFactory(as4CryptoProperties)
                .messageProperties(getMandatoryMessageProperties(message.metadata().getMessageProperties()))
                .messageID(UUID.randomUUID().toString())
                .action(message.metadata().getCollaborationInfo().action())
                .service(message.metadata().getCollaborationInfo().service())
                .senderParticipantID(aSenderID)
                .receiverParticipantID(aReceiverID)
                .fromPartyID(aSenderID)
                .fromRole(message.metadata().getPartyInfo().fromRole())
                .toPartyID(aReceiverID)
                .toRole(message.metadata().getPartyInfo().toRole())
                .receiverEndpointDetails(as4CryptoProperties.getCertificateForAlias(message.metadata().getPartyInfo().toPartyId()), as4EndpointProperties.getSendDestination())
                .httpRetrySettings(new HttpRetrySettings().setMaxRetries(0));

        CryptoSettings.apply(builder);

        message.payload().ifPresent(builder::payload);

        message.attachments().forEach(builder::addAttachment);

        return builder;
    }
```
  2. `CryptoSettings` is from my post above.
  and `as4CryptoProperties` below:
```java
public class AS4CryptoProperties extends AbstractAS4CryptoFactory {

    private final String CRYPTO_PROVIDER;
    private final String KEYSTORE_TYPE;
    private final String KEYSTORE_FILE;
    private final String KEYSTORE_PASSWORD;
    private final String KEY_ALIAS;
    private final String KEY_PASSWORD;
    private final String LOAD_CACERTS;
    private final String TRUSTSTORE_FILE;
    private final String TRUSTSTORE_PASSWORD;
    private final String TRUSTSTORE_PROVIDER;
    private final String TRUSTSTORE_TYPE;

    @Getter
    @NotNull
    private final X509Certificate certificate;

    private final Properties properties;

    @Autowired
    public AS4CryptoProperties(Environment env) {
        final String APP_HOME = Optional.fromNullable(env.getProperty("mock.app.home")).or(Strings.EMPTY);
        if (Strings.isNotEmpty(APP_HOME)) {
            KEYSTORE_FILE = APP_HOME + "/" + env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_FILE);
            TRUSTSTORE_FILE = APP_HOME + "/" + env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_FILE);
        } else {
            KEYSTORE_FILE = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_FILE);
            TRUSTSTORE_FILE = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_FILE);
        }
        CRYPTO_PROVIDER = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.CRYPTO_PROVIDER);
        KEYSTORE_TYPE = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_TYPE);
        KEYSTORE_PASSWORD = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_PASSWORD);
        KEY_ALIAS = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEY_ALIAS);
        KEY_PASSWORD = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.KEY_PASSWORD);
        LOAD_CACERTS = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.LOAD_CACERTS);
        TRUSTSTORE_PASSWORD = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_PASSWORD);


        log.debug(this.toString());

        TRUSTSTORE_PROVIDER = Optional.fromNullable(env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_PROVIDER)).orNull();
        TRUSTSTORE_TYPE = env.getProperty(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_TYPE);

        KeyStoreHelper.setResourceProvider(new DefaultResourceProvider());

        m_aKeyStore = KeyStoreHelper.loadKeyStore(Objects.requireNonNull(EKeyStoreType.getFromIDCaseInsensitiveOrDefault(KEYSTORE_TYPE, DEFAULT_KEYSTORE_TYPE)),
                        KEYSTORE_FILE,
                        KEYSTORE_PASSWORD)
                .getKeyStore();

        m_aTrustStore = KeyStoreHelper.loadKeyStore(Objects.requireNonNull(EKeyStoreType.getFromIDCaseInsensitiveOrDefault(TRUSTSTORE_TYPE, DEFAULT_TRUSTSTORE_TYPE)),
                        TRUSTSTORE_FILE,
                        TRUSTSTORE_PASSWORD)
                .getKeyStore();

        properties = new Properties();
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.CRYPTO_PROVIDER, CRYPTO_PROVIDER);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.KEY_ALIAS, KEY_ALIAS);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.KEY_PASSWORD, KEY_PASSWORD);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_FILE, KEYSTORE_FILE);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_PASSWORD, KEYSTORE_PASSWORD);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.KEYSTORE_TYPE, KEYSTORE_TYPE);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.LOAD_CACERTS, LOAD_CACERTS);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_FILE, TRUSTSTORE_FILE);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_PASSWORD, TRUSTSTORE_PASSWORD);
        properties.put(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_TYPE, TRUSTSTORE_TYPE);

        if (Objects.nonNull(TRUSTSTORE_PROVIDER)) {
            properties.put(com.helger.phase4.crypto.AS4CryptoProperties.TRUSTSTORE_PROVIDER, TRUSTSTORE_PROVIDER);
        }

        CryptoType cryptoType = new CryptoType(CryptoType.TYPE.ALIAS);
        cryptoType.setAlias(KEY_ALIAS);
        try {
            certificate = Arrays.stream(getCrypto().getX509Certificates(cryptoType)).findFirst().orElseThrow(() -> new RuntimeException("Couldn't get X509 certificate based on provided KEY_ALIAS and KEYSTORE"));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private final KeyStore m_aKeyStore;
    private final KeyStore m_aTrustStore;

    public X509Certificate getCertificateForAlias(String alias) {
        CryptoType cryptoType = new CryptoType(CryptoType.TYPE.ALIAS);
        cryptoType.setAlias(alias);
        try {
            return Arrays.stream(getCrypto().getX509Certificates(cryptoType)).findFirst().orElseThrow(() -> new RuntimeException("Couldn't get X509 certificate based on provided alias: " + alias));
        } catch (WSSecurityException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Lazily create a {@link Crypto} instance using the properties from
     * {@link #()}.
     */
    @Nonnull
    public final Crypto getCrypto(@Nonnull final ECryptoMode eCryptoMode) {
        return getCrypto();
    }

    private Crypto getCrypto() {
        try {
            return CryptoFactory.getInstance(properties);
        } catch (WSSecurityException e) {
            throw new RuntimeException(e);
        }
    }

    @Nullable
    public final KeyStore getKeyStore() {
        return m_aKeyStore;
    }

    @Nullable
    public final String getKeyAlias() {
        return KEY_ALIAS;
    }

    @Nullable
    public final String getKeyPassword() {
        return KEY_PASSWORD;
    }

    @Nullable
    public String getKeyPasswordPerAlias(@Nullable final String sSearchKeyAlias) {
        final String sKeyAlias = getKeyAlias();

        // Use case insensitive compare, depends on the keystore type
        if (sKeyAlias != null && sKeyAlias.equalsIgnoreCase(sSearchKeyAlias))
            return KEY_PASSWORD;

        return null;
    }

    @Nullable
    public final KeyStore getTrustStore() {
        return m_aTrustStore;
    }
}
```

    
  3. When sending message to Domibus with leg config of policy config:
```xml
              <sp:AsymmetricBinding>
                <wsp:Policy>
                    <sp:InitiatorToken>
                        <wsp:Policy>
                            <sp:X509Token
                                    sp:IncludeToken="http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702/IncludeToken/Always">
                                <wsp:Policy>
                                    <sp:WssX509V3Token10/>
                                </wsp:Policy>
                            </sp:X509Token>
                        </wsp:Policy>
                    </sp:InitiatorToken>
                    <sp:RecipientToken>
                        <wsp:Policy>
                            <sp:X509Token
                                    sp:IncludeToken="http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702/IncludeToken/Always">
                                <wsp:Policy>
                                    <sp:WssX509V3Token10/>
                                </wsp:Policy>
                            </sp:X509Token>
                        </wsp:Policy>
                    </sp:RecipientToken>
                    <sp:Layout>
                        <wsp:Policy>
                            <sp:Strict/>
                        </wsp:Policy>
                    </sp:Layout>
                    <sp:SignBeforeEncrypting/>
                    <sp:AlgorithmSuite>
                        <wsp:Policy>
                            <sp-cxf:Basic128GCMSha256MgfSha256
                                    xmlns:sp-cxf="http://e-delivery.eu/custom/security-policy"/>
                        </wsp:Policy>
                    </sp:AlgorithmSuite>
                </wsp:Policy>
            </sp:AsymmetricBinding>
```
  4. Result in exeption when Domibus is trying decrypt message `javax.crypto.AEADBadTagException: mac check in GCM failed`

**Reply by phax** [reply]
Well, my proposal would be to skip `CryptoSettings.apply(builder);` - all of these things should be predefined in the PMode

**Reply by oaxfreeman** [reply]
Ok, but I do not understand how do you actually configure PMode in Phase4? rr more precisely, how to configure PModeLegSecurity from any Sender class? I can see that profiles hardcode some crypto settings but how to override them?

**Reply by phax** [reply]
Well, you include the `phase4-profile-cef` module in your POM. This will trigger the registration of the CEF AS4 profiles via Java SPI technology.
By using the class `Phase4CEFSender.CEFUserMessageBuilder` you implicitly choose to use the CEF AS4 Profile (4 corners) which in turn make sure that the `CEFPMode` is used.
So all the security is abstracted away from you by default. You should only deal with the security parameters, if you need something special or something that cannot be delivered out of the box (e.g. for the BDEW profile)

**Reply by oaxfreeman** [reply]
I'm working with Phase4 `2.9.0`
So far I've managed to :
- register "Bouncy Castle" as `provider[0]` within Phase4
- enure same version of "Bouncy Castle" between domibus and phase4
- checked `Cipher` properties, padding, nonce etc.. 

everything is same, still getting `javax.crypto.AEADBadTagException: mac check in GCM failed`

So I've decided to upgrade to `3.0.1`
Now I struggle to recreate :
 `Phase4CEFSender .receiverEndpointDetails(as4CryptoProperties.getCertificateForAlias(message.metadata().getPartyInfo().toPartyId()), as4EndpointProperties.getSendDestination())` since as4CryptoProperties of `AbstractAS4CryptoFactory` is deprecated?

**Reply by phax** [reply]
There should not be any difference between 2.9.0 and 3.0.1 as the security components haven't changed.
Some short questions on what you managed so far:
* The registration of BC as a provider should not be neceesary. Can you please paste the code that you are using?
* What version of BC are we talking about? To my understanding, the main encryption part should be done by the Java runtime and not BC
* Do you have your "CryptoSettings.apply" still in use?

Regarding the migration to v3.0.0. The properties based approach is a bit redundant, so I suggest you use `AS4CryptoFactoryInMemoryKeyStore` and pass in KeyStore and TrustStore as a parameter.

Eventually, if you write me an email (see pom.xml) I will invite you to my Slack channel and we can do a huddle with screensharing.

**Reply by oaxfreeman** [reply]
At the moment we are working with `Phase4:3.0.1`
`BC` is not registered within `Security.insertProvider` and it is not present in `pom.xml`
`CryptoSettings.appy` is not in use
So we can say we send with `default` settings to domibus which expects AES-128-GCM, still getting `mac check in GCM failed`  :-D
Also I was forced to add `ph-commons` to pom.xml since classes from `3.0.1` weren't resolved

**Reply by oaxfreeman** [reply]
@phax I'm embarrassed to let you know that it is a problem with trustore certificate used in communication with Domibus, and unfortunately `Domibus` is throwing Error `mac check in GCM failed` wich is very missleading in this case..

**Reply by phax** [reply]
I am glad you were able to resolve the issue :)

**Reply by oaxfreeman** [reply]
Nonetheless `Phase4` is proven to work with `Domibus` using `AES-128 GCM` Cipher - without any additional crypto configuration
